drivers/sensor/melexis/MLX90394/mlx90394.c

```diff
@@ -0,0 +1,601 @@
+{
+	int64_t scale;
+
+	if (config == MLX90394_CTRL2_CONFIG_HIGH_SENSITIVITY_LOW_NOISE) {
+		scale = MLX90394_HIGH_SENSITIVITY_MICRO_GAUSS_PER_BIT;
+	} else {
+		scale = MLX90394_HIGH_RANGE_MICRO_GAUSS_PER_BIT;
+	}
+	int64_t conv_val = (int16_t)((uint16_t)sample_l | (uint16_t)(sample_h << 8)) * scale;
+
+	val->val1 = conv_val / 1000000;               /* G */
+	val->val2 = conv_val - (val->val1 * 1000000); /* uG */
+}
+
+static void mlx90394_convert_temp(struct sensor_value *val, uint8_t sample_l, uint8_t sample_h)
+{
+	int64_t conv_val =
+		sys_le16_to_cpu(sample_l | (sample_h << 8)) * MLX90394_MICRO_CELSIUS_PER_BIT;
+	val->val1 = conv_val / 1000000;                 /* C */
+	val->val2 = (conv_val - (val->val1 * 1000000)); /* uC */
+}
+
+/*
+ * The user has to take care about that the requested channel was fetched before. Else the data will
+ * be random. Magnetic Flux Density is in Gauss, Temperature in Celsius
+ */
+static int mlx90394_channel_get(const struct device *dev, enum sensor_channel chan,
+				struct sensor_value *val)
+{
+	struct mlx90394_data *data = dev->data;
+
+	switch (chan) {
+	case SENSOR_CHAN_MAGN_X: {
+		mlx90394_convert_magn(data->config_val, val, data->sample.x_l, data->sample.x_h);
+	} break;
+	case SENSOR_CHAN_MAGN_Y: {
+		mlx90394_convert_magn(data->config_val, val, data->sample.y_l, data->sample.y_h);
+	} break;
+	case SENSOR_CHAN_MAGN_Z: {
+		mlx90394_convert_magn(data->config_val, val, data->sample.z_l, data->sample.z_h);
+	} break;
+	case SENSOR_CHAN_AMBIENT_TEMP: {
+		mlx90394_convert_temp(val, data->sample.temp_l, data->sample.temp_h);
+	}
+	case SENSOR_CHAN_MAGN_XYZ: {
+		mlx90394_convert_magn(data->config_val, val, data->sample.x_l, data->sample.x_h);
+		mlx90394_convert_magn(data->config_val, val + 1, data->sample.y_l,
+				      data->sample.y_h);
+		mlx90394_convert_magn(data->config_val, val + 2, data->sample.z_l,
+				      data->sample.z_h);
+	} break;
+	case SENSOR_CHAN_ALL: {
+		mlx90394_convert_magn(data->config_val, val, data->sample.x_l, data->sample.x_h);
+		mlx90394_convert_magn(data->config_val, val + 1, data->sample.y_l,
+				      data->sample.y_h);
+		mlx90394_convert_magn(data->config_val, val + 2, data->sample.z_l,
+				      data->sample.z_h);
+		mlx90394_convert_temp(val + 3, data->sample.temp_l, data->sample.temp_h);
+	}
+	default: {
+		LOG_DBG("Invalid channel %d", chan);
+		return -ENOTSUP;
+	}
+	}
+	return 0;
+}
+
+static inline int mlx90394_update_register(const struct device *dev, const uint8_t reg_addr,
+					   const uint8_t new_val, uint8_t *old_value)
+{
+	const struct mlx90394_config *cfg = dev->config;
+
+	if (new_val != *old_value) {
+		*old_value = new_val;
+		return i2c_reg_write_byte_dt(&cfg->i2c, reg_addr, new_val);
+	} else {
+		return 0;
+	}
+}
+
+static inline int mlx90394_fs_set(const struct device *dev, const struct sensor_value *val)
+{
+	struct mlx90394_data *data = dev->data;
+
+	if (data->config_val > MLX90394_CTRL2_CONFIG_HIGH_RANGE_LOW_NOISE) {
```

Comment:
Prefer to check the error case first to do an early return, and save indentation space.

```suggestion
	if (data->config_val <= MLX90394_CTRL2_CONFIG_HIGH_RANGE_LOW_NOISE) {
		LOG_ERR("different FS values only supported in low noise mode");
		return -ENOTSUP;
	}
	
	// ...
```

Reply:
accepted